actionpack/lib/action_dispatch/routing/url_for.rb

```diff
@@ -171,6 +171,10 @@ def url_for(options = nil)
           route_name = options.delete :use_route
           _routes.url_for(options.symbolize_keys.reverse_merge!(url_options),
                          route_name)
+        when ActionController::Parameters
```

Comment:
Sounds good. There actually are tests that failed if I didn't modify this line, though.

I'll let you decide if you want to get #20797 in first and have me rebase it or not.